src/ipips/ipip-0412.md

```diff
@@ -0,0 +1,260 @@
+   read and returned to the consumer) and when it's received might not match.
+1. Blocks can be reused multiple times, this is handy for cases when you plan
+   to cache on disk but not at all when you want to process a stream with use &
+   forget policy.
+
+What we really want is for the gateway to help us a bit, and give us blocks in
+a useful order.
+
+The existing Trustless Gateway specification does not provide a mechanism for
+negotiating the order of blocks in CAR responses.
+
+This IPIP aims to improve the status quo.
+
+## Detailed design
+
+CAR content type
+([`application/vnd.ipld.car`](https://www.iana.org/assignments/media-types/application/vnd.ipld.car))
+already supports `version` parameter, which allows gateway to indicate which
+CAR flavour is returned with the response.
+
+The proposed solution introduces two new parameters for the content type headers
+in HTTP requests and responses: `order` and `dups`.
+
+The `order` parameter lets the client specify the desired block order in the
+CAR response, while the `dups` parameter determines if a block is sent multiple
+times when it appears more than once in the requested DAG.
+
+### Signaling in Request
+
+Content type negotiation is based on section 12.5.1 of :cite[rfc9110].
+
+Clients MAY indicate their preferred block order by sending an `Accept` header in
+the HTTP request. The `Accept` header format is as follows:
+
+```
+Accept: application/vnd.ipld.car; version=1; order=dfs; dups=y
+```
+
+In the future, when more orders or parameters exist, clients will be able to
+specify a list of preferences, for example:
+
+```
+Accept: application/vnd.ipld.car;order=foo, application/vnd.ipld.car;order=dfs;dups=y;q=0.5
+```
+
+The above example is a list of preferences, the client would really like to use
+the hypothetical `order=foo` however if this isn't available it would accept
+`order=dfs` with `dups=y` instead (lower priority indicated via `q` parameter,
+as noted in :cite[rfc9110]).
+
+#### `order` CAR content type parameter
+
+The `order` parameter allows clients to specify the desired block order in the
+response. It supports the following values:
+
+- `dfs`: [Depth-First Search](https://en.wikipedia.org/wiki/Depth-first_search)
+  order, enables streaming responses with minimal memory usage.
+- `unk`: Unknown order, which serves as the implicit default when the order
+  parameter is missing. In this case, the client cannot make any assumptions
+  about the block order: blocks may arrive in a random order or be a result of
+  a custom DAG traversal algorithm.
+
+#### `dups` CAR content type parameter
+
+The `dups` parameter specifies whether duplicate blocks (the same block
+occuring multiple times in the requested DAG) will be present in the CAR
+response. Useful when a deterministic block order is used.
+
+It accepts two values:
+- `y`: Duplicate blocks MUST be sent every time they occur during the DAG walk.
+- `n`: Duplicate blocks MUST be sent only once.
```

Comment:
Using a thread because IMO makes things a little easier to manage since there might be some opinions here 🙃 , but think @olizilla has some good points here https://github.com/ipfs/specs/pull/412#pullrequestreview-1430979342.

`dups=n` with a `MUST` can be pretty expensive to implement with certain scenarios (as can `dups=y`, although for UnixFS data it's not really more expensive than implementing the trusted gateway equivalent behavior).

Are the people who need `dups=n` in need of a MUST here, or is something like returning at most the duplicate blocks from `y` ok? This does make life a little harder for clients who would otherwise just be writing the data to disk as they now have to do a little more work and guessing to determine if getting data from someone who effectively sends them `dups=y` labelled as `dups=n` is worth dealing with in the future.

From my perspective no strong feelings on multiple parameters vs one name for all the things. If people really wanted strong etags and the ability to export CARs in a very particular way (e.g. matching how certain Filecoin imports used to work) then having names for the combination of parameters seems like a good idea. IIUC that's not the case though so 🤷.

Reply:
The meaning behind the presence of `dups=y|n` is MUST: you signal need or that you returned a deterministic thing.
that is a feature, because content-type parameters are optional,  so if you  dont know, dont care, or don't want to specify, you don't include it.

Order and duplicates are different properties, but clumping various things like thaty together under `pack` concept proposed by @olizilla (`pack=dfs` and `pack=dfs-stream`) makes it easier to test and reason about.

So we would have:
- `order=dfs` (unknown `dupes`) →  `pack=dfs`
- `order=dfs; dupes=y` → `pack=dfs-stream`

and we would have to only care about testing these specific bundles + easier to add future `pack` strategies without balooning parameter space.

cc @hannahhoward @fabricedesre thoughts?

Reply:
I've been thinking a bit about this the last few days as I'm working on some implementation things and I think what I'd like is something like this - being able to _request_ that a particular order be used, but then being _told_ what order is being used in the response. I'd like to be able to default to the more network-efficient `dups=n` but choose to switch to `dups=y` where a user doesn't want to burn temporary and/or wants to cleanly stream; but knowing that there are cases where `dups=n` isn't going to be practical (eipfs notably), I'd like to be told that by the server in order to make decisions on the client end - I can proceed and switch modes internally, or maybe bail because it doesn't fit within client constraints.

The same could be said for `order`. Lassie will always (currently) request `order=dfs` but if we get back an `order=something else` then we could either abort early, or choose to handle it in some other way (maybe the client trusts the provider explicitly and doesn't need the implicit guarantees of trustless dfs traversal so random is fine).

Could we add these to the `Content-Type` response header? `application/vnd.ipld.car; version=1; dups=y; order=dfs`.

Reply:
Also worth noting here that this is possible in CARv2, we could choose to go that route, it's slightly more annoying though. `application/vnd.ipld.car; version=2`, then use an indexless carv2 with characteristic bits set for order and dups.

Reply:
@rvagg  yes, this is ok. Including `version` in response's `Content-Type`  is already happening in kubo/boxo/gateway/production sinc last year.

This IPIP follows that convention, and states we want gateway to return `dups` and `order` too (if these are known, even when not explicitly asked for):

> ### Signaling in Response
> 
> The Trustless Gateway MUST always respond with a `Content-Type` header that includes
> information about all supported/known parameters, even if the client did not
> specify them in the request.
> 
> The `Content-Type` header format is as follows:
> 
> ```
> Content-Type: application/vnd.ipld.car;version=1;order=dfs;dups=y
> ```

Reply:
@rvagg so is the remaining ask here to 

- [ ] add an explicit spec note that "Gateway MAY choose to return different order/dups than the requested one instead of just erroring"

? 

It makes sense imo, because that is what will effectively be already happening in production with legacy gateways that do not support IPIP-412.